parsec/backend/asgi/rpc.py

```diff
@@ -268,3 +281,138 @@ async def authenticated_api(raw_organization_id: str) -> Response:
+        self._body_line_sender, self._body_line_receiver = trio.open_memory_channel[bytes](0)
+        self._contextmanager = self._make_contextmanager(client_ctx, backend)
+
+    @asynccontextmanager
+    async def _make_contextmanager(
+        self, client_ctx: AuthenticatedClientContext, backend: BackendApp
+    ) -> AsyncIterator[None]:
+        async def _events_into_body_lines() -> None:
+            try:
+                # In SSE, the HTTP status code & headers are sent with the first event.
+                # This means the client has to wait for this first event to know for
+                # sure the connection was successful (in practice the server responds
+                # fast in case of error and potentially up to the keepalive time in case
+                # everything is ok).
+                # While not strictly needed, we send a keepalive event (i.e. an event
+                # with no name and any comment, see https://html.spec.whatwg.org/multipage/server-sent-events.html#authoring-notes
+                # ) right away so that the client knows it is correctly connected without delay.
+                await self._body_line_sender.send(b":keepalive\n\n")
+
+                if backend.config.sse_keepalive is None:
+                    async for event in client_ctx.receive_events_channel:
+                        body_line = b"data:" + b64encode(event.dump()) + b"\n\n"
+                        await self._body_line_sender.send(body_line)
+
+                else:
+                    while True:
+                        with trio.move_on_after(backend.config.sse_keepalive) as scope:
+                            try:
+                                print("~~~~~wait event")
+                                event = await client_ctx.receive_events_channel.receive()
+                                print("~~~~~event", event)
+                            except trio.EndOfChannel:
+                                break
+                            body_line = b"data:" + b64encode(event.dump()) + b"\n\n"
+                            await self._body_line_sender.send(body_line)
+
+                        if scope.cancelled_caught:
+                            body_line = b":keepalive\n\n"
+                            await self._body_line_sender.send(body_line)
+
+            finally:
+                # Closing sender end of the channel will cause Quart stop iterating
+                # on the receiver end and hence terminate the request
+                self._body_line_sender.close()
+
+        assert client_ctx.cancel_scope is not None
+        with client_ctx.cancel_scope:
+            async with trio.open_nursery() as nursery:
+
+                with backend.event_bus.connection_context() as client_ctx.event_bus_ctx:
+                    await backend.events.connect_events(client_ctx)
+
+                    nursery.start_soon(_events_into_body_lines)
+
+                    yield
```

Comment:
I guess it would make sense to also use a context manager for `self._body_line_receiver` around the `yield` statement.

Reply:
I didn't add it the context manager on `self._body_line_receiver` so that `self._body_line_sender.send` will never raise a `BrokenResourceError`.
The idea being the cancellation propagate from producer to consumer (i.e. the receiver code is quart doing an async for loop forever)

However after closer investigation, there was an issue in the code if Quart decides (e.g. when server closes) to cancel the query. I've updated the code and added some looong comments ^^